ibtk/src/lagrangian/fortran/lagrangian_interaction2d.f.m4

```diff
@@ -3187,3 +3187,486 @@ c
+c
+      subroutine lagrangian_composite_bspline_32_interp2d(
+     &     dx,x_lower,x_upper,depth,
+     &     ilower0,iupper0,ilower1,iupper1,
+     &     nugc0,nugc1,
+     &     u,
+     &     indices,Xshift,nindices,
+     &     X,V,
+     &     axis)
+c
+      implicit none
+c
+c     Functions.
+c
+      REAL lagrangian_piecewise_linear_delta
+      REAL lagrangian_bspline_3_delta
+c
+c     Input.
+c
+      INTEGER depth
+      INTEGER ilower0,iupper0,ilower1,iupper1
+      INTEGER nugc0,nugc1
+      INTEGER nindices
+      INTEGER axis
+
+      INTEGER indices(0:nindices-1)
+
+      REAL Xshift(0:NDIM-1,0:nindices-1)
+
+      REAL dx(0:NDIM-1),x_lower(0:NDIM-1),x_upper(0:NDIM-1)
+      REAL u(CELL2dVECG(ilower,iupper,nugc),0:depth-1)
+      REAL X(0:NDIM-1,0:*)
+c
+c     Input/Output.
+c
+      REAL V(0:depth-1,0:*)
+c
+c     Local variables.
+c
+      INTEGER ilower(0:NDIM-1),iupper(0:NDIM-1)
+      INTEGER ic,ic0,ic1
+      INTEGER ic_center(0:NDIM-1),ic_lower(0:NDIM-1),ic_upper(0:NDIM-1)
+      INTEGER d,l,s,nugc(0:NDIM-1)
+
+      REAL X_cell(0:NDIM-1),w(0:NDIM-1,0:2)
+c
+c     Prevent compiler warning about unused variables.
+c
+      x_upper(0) = x_upper(0)
+c
+c     Setup convenience arrays.
+c
+      ilower(0) = ilower0
+      ilower(1) = ilower1
+
+      iupper(0) = iupper0
+      iupper(1) = iupper1
+
+      nugc(0) = nugc0
+      nugc(1) = nugc1
+c
+c     Use a composite B-spline function to interpolate u onto V.
+c
+      do l = 0,nindices-1
+         s = indices(l)
+         do d=0,NDIM-1
+c
+c     Determine the Cartesian cell in which X(s) is located.
+c
+            ic_center(d) =
+     &           floor((X(d,s)+Xshift(d,l)-x_lower(d))/dx(d))
+     &           + ilower(d)
+            X_cell(d) = x_lower(d)
+     &           +(dble(ic_center(d)-ilower(d))+0.5d0)*dx(d)
+c
+c     Determine the interpolation stencil corresponding to the position
+c     of X(s) within the cell.
+c
+            ic_lower(d) = ic_center(d)-1
+            ic_upper(d) = ic_center(d)+1
+            ic_lower(d) = max(ic_lower(d),ilower(d)-nugc(d))
+            ic_upper(d) = min(ic_upper(d),iupper(d)+nugc(d))
+c
+c     Compute the interpolation weights.
+c
+            do ic = ic_lower(d),ic_upper(d)
+               X_cell(d) = x_lower(d)+(dble(ic-ilower(d))+0.5d0)*dx(d)
+               if (d .eq. axis) then
+                  w(d,ic-ic_lower(d)) =
+     &                 lagrangian_bspline_3_delta(
+     &                 (X(d,s)+Xshift(d,l)-X_cell(d))/dx(d))
+               else
+                  w(d,ic-ic_lower(d)) =
+     &                 lagrangian_piecewise_linear_delta(
+     &                 (X(d,s)+Xshift(d,l)-X_cell(d))/dx(d))
+               endif
+            enddo
+         enddo
+c
+c     Interpolate u onto V.
+c
+         INTERPOLATE_2D_SPECIALIZE_FIXED_WIDTH(ic_lower(1), ic_upper(1),
+                                               ic_lower(0), ic_upper(0),
+                                               3)
```

Comment:
if we are reading a 3x3 stencil here then we need to populate the extra rows (or columns) of `w` with zeros.

Reply:
That happens automatically because the basic kernel functions all return 0 outside of their radius of support.

Reply:
However, per #1501, we could probably improve performance by making specialized helper macros that only use the nonzero values. For staggered data, we also probably also could relax the ghost cell width requirements for these kernels.